### ISSUE TITLE
Psalm unexpectedly narrows return types of closures

### ISSUE THREAD

**Issue by weirdan** [vimeo/psalm#2089]
https://psalm.dev/r/e58cebdc53
```php
<?php
  
function acceptsSome(Some $_m): void {}

/** @template T */
interface Maybe {}

/**
 * @template T
 * @template-implements Maybe<T>
 */
class Some implements Maybe
{
    /** @psalm-param T $_value */
    public function __construct($_value) {}
}

$maybe =
  /** @return Maybe<'foo'> */
  function(): Maybe {
	return new Some('foo');
  }
;
acceptsSome($maybe()); // Should complain about ArgumentTypeCoercion, but doesn't

/** @return Maybe<'foo'> */
function maybe(): Maybe {
  return new Some('foo');
};
acceptsSome(maybe()); // Complains as expected
```
**Expected**: both calls to `acceptsSome()` are flagged, as both get passed `Maybe`
**Actual**: Psalm figures out that `$maybe` always returns `Some`:
```
Psalm output (using commit 504ccda): 

ERROR: ArgumentTypeCoercion - 30:13 - Argument 1 of acceptsSome expects Some, parent type Maybe<string(foo)> provided
```

I thinks it's fine to infer the return type when the closure has none specified explicitly, but when it has the docblock return type or the return type hint I wouldn't expect Psalm to ignore (or narrow) it.

**Comment by muglug** [comment]
The example can be simplified to 

```php
class A {}
class B extends A {}

function takesA(A $_a) : void {}
function takesB(B $_b) : void {}

$getAButReallyB =
/** @return A */
function(): A {
  return new B;
};

takesA($getAButReallyB());
takesB($getAButReallyB()); // this is fine

function getAButReallyB(): A {
  return new B;
};

takesA(getAButReallyB());
takesB(getAButReallyB()); // errors
```

https://psalm.dev/r/dc30f60a36

This was deliberate, as closures can generally have a more accurate return type inferred inline. But I agree that if there's a docblock already present on the closure, Psalm should use that.
